### ISSUE TITLE
TypeError: __init__() got an unexpected keyword argument 'differentiable'

### ISSUE THREAD

**Issue by comaniac** [huggingface/accelerate#801]
### System Info

```Shell
- `Accelerate` version: 0.13.0.dev0 (commit 6dc429f6f748600c4319e438f6bdaa22261c4f6a)
- Platform: Linux-5.4.0-1075-aws-x86_64-with-glibc2.17
- Python version: 3.8.13
- Numpy version: 1.23.4
- PyTorch version (GPU?): 1.13.0+cu117 (True)
- `Accelerate` default config:
        - compute_environment: LOCAL_MACHINE
        - deepspeed_config: {}
        - distributed_type: FSDP
        - downcast_bf16: no
        - fsdp_config:
          - fsdp_auto_wrap_policy: TRANSFORMER_BASED_WRAP
          - fsdp_backward_prefetch_policy: NO_PREFETCH
          - fsdp_offload_params: false
          - fsdp_sharding_strategy: 1
          - fsdp_state_dict_type: FULL_STATE_DICT
          - fsdp_transformer_layer_cls_to_wrap: GPT2Block
        - machine_rank: 0
        - main_process_ip: null
        - main_process_port: null
        - main_training_function: main
        - mixed_precision: fp16
        - num_machines: 1
        - num_processes: 8
        - rdzv_backend: static
        - same_network: true
        - use_cpu: false
```

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [X] One of the scripts in the examples/ folder of Accelerate or an officially supported `no_trainer` script in the `examples` folder of the `transformers` repo (such as `run_no_trainer_glue.py`)
- [ ] My own task or dataset (give details below)

### Reproduction

```
cd transformers; accelerate launch transformers/examples/pytorch/language-modeling/run_clm_no_trainer.py \
--dataset_name wikitext --dataset_config_name wikitext-2-raw-v1 --block_size 1024 --overwrite_cache \
--output_dir /tmp/gpt --model_name_or_path gpt2-medium --tokenizer_name gpt2-medium \
--per_device_train_batch_size 1 --max_train_steps 20 --learning_rate 2e-5 --gradient_accumulation_steps 1
```

### Expected behavior

```Shell
The example should run without issues. However, this is what I got:


  File "transformers/examples/pytorch/language-modeling/run_clm_no_trainer.py", line 676, in <module>
    main()
  File "transformers/examples/pytorch/language-modeling/run_clm_no_trainer.py", line 496, in main
    model, optimizer, train_dataloader, eval_dataloader, lr_scheduler = accelerator.prepare(
  File "/home/ubuntu/workspace_hf/accelerate/src/accelerate/accelerator.py", line 673, in prepare
    result = self._prepare_fsdp(*result)
  File "/home/ubuntu/workspace_hf/accelerate/src/accelerate/accelerator.py", line 585, in _prepare_fsdp
    optimizer = obj.optimizer.__class__(model.parameters(), **obj.optimizer.defaults)
TypeError: __init__() got an unexpected keyword argument 'differentiable'
```

It is likely due to this PyTorch commit: https://github.com/pytorch/pytorch/pull/86183. After this commit, Adam series optimizers have `differentiable` as one of the default arguments, but seems like the real optimizer constructors do not have this argument.

I could make this example work by adding two lines:

```python
if "differentiable" in obj.optimizer.defaults:
    del obj.optimizer.defaults["differentiable"]
optimizer = obj.optimizer.__class__(model.parameters(), **obj.optimizer.defaults)
```
```

**Comment by sgugger** [comment]
Thanks for flagging! I have no idea why the defaults are no longer in sync with the kwargs accepted by the optimizer, but your fix looks good. Please go ahead with a PR!

**Comment by comaniac** [comment]
I dived into PyTorch and kind of getting some senses:

1. PR https://github.com/pytorch/pytorch/pull/80938 in 3 months ago introduces `differentiable` to `defaults`.
2. PR https://github.com/pytorch/pytorch/pull/86183 added `differentiable` to `__init__`, which should makes it work.

However, the PyTorch 1.13.0 release at the commit: https://github.com/pytorch/pytorch/tree/7c98e70d44abc7a1aead68b6ea6c8adc8c554db5, which doesn't include https://github.com/pytorch/pytorch/pull/86183.

In summary, with PyTorch 1.13.0, **calling `__setstate__`** to recover the optimizer will bring `differentiable` to `defaults`, but using this `defaults` to initialize another optimizer without https://github.com/pytorch/pytorch/pull/86183 would crash.

To make `accelerate` work with PyTorch 1.13.0 and future versions, I'll go ahead and file a PR to use a try-catch block to remove `differentiable` if needed.

**Comment by Philmod** [comment]
I'm still seeing this error with the version `0.20.3` https://github.com/Kaggle/docker-python/pull/1271

Is this expected?
